src/plugins/obj/obj_backend.cpp

```diff
@@ -18,12 +18,65 @@
 #include "obj_backend.h"
 #include "common/nixl_log.h"
 #include "nixl_types.h"
+#include <absl/strings/str_format.h>
 #include <memory>
+#include <future>
+#include <vector>
+#include <chrono>
+
+bool
+isValidPrepXferParams (const nixl_xfer_op_t &operation,
+                      const nixl_meta_dlist_t &local,
+                      const nixl_meta_dlist_t &remote,
+                      const std::string &remote_agent,
+                      const std::string &local_agent) {
+    if (remote_agent != local_agent) {
+        NIXL_ERROR << absl::StrFormat("Error: Remote agent must match the requesting agent (%s). Got %s",
+                                    local_agent, remote_agent);
+        return false;
+    }
+
+    if (local.getType() != DRAM_SEG) {
+        NIXL_ERROR << absl::StrFormat("Error: Local memory type must be DRAM_SEG, got %d", local.getType());
+        return false;
+    }
+
+    if (remote.getType() != OBJ_SEG) {
+        NIXL_ERROR << absl::StrFormat("Error: Remote memory type must be OBJ_SEG, got %d", remote.getType());
+        return false;
+    }
+
+    if (local.descCount() != remote.descCount()) {
+        NIXL_ERROR << absl::StrFormat("Error: Mismatch in descriptor counts - local: %d, remote: %d",
+                                    local.descCount(), remote.descCount());
+        return false;
+    }
+
+    return true;
+}
 
 class nixlObjBackendReqH : public nixlBackendReqH {
 public:
-    std::atomic_bool completed_ = false;
-    std::atomic<nixl_status_t> status_;
+    nixlObjBackendReqH() = default;
+    ~nixlObjBackendReqH() = default;
+
+    std::vector<std::shared_future<nixl_status_t>> status_futures_;
+
+    nixl_status_t getOverallStatus() {
+        nixl_status_t status = NIXL_SUCCESS;
+
+        for (auto& future : status_futures_) {
+            if (future.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
+                auto current_status = future.get();
```

Comment:
Traversing whole vector of futures every single time user checks the status is suboptimal. How about starting from the end of the vector and pop every completed future so that it would only be checked once? On error return the error code and clear the whole vector.